File: game_plugins/hase_und_igel_new/shared/sc/plugin2018/Advance.java
package sc.plugin2018;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;
import sc.plugin2018.util.GameRuleLogic;
import sc.shared.InvalidMoveException;

/** TODO comment everything
 * Ein Vorwärtszug, um spezifizierte Distanz. Verbrauchte Karroten werden mit k = (distance * (distance + 1)) / 2
 * berechnet (Gaußsche Summe)
 */
@XStreamAlias(value = "advance")
public class Advance extends Action{

  @XStreamAsAttribute
  private int distance;

  public Advance(int distance) {
    super();
    setDistance(distance);
  }

  public Advance(int distance, int order) {
    this.order = order;
    setDistance(distance);
  }


  @Override
  public void perform(GameState state) throws InvalidMoveException {
    if (GameRuleLogic.isValidToAdvance(state, this.distance)) {
      state.getCurrentPlayer().changeCarrotsAvailableBy(- GameRuleLogic.calculateCarrots(this.distance));
      state.getCurrentPlayer().setFieldIndex(state.getCurrentPlayer().getFieldIndex() + distance);
      if (state.getTypeAt(state.getCurrentPlayer().getFieldIndex()) == FieldType.HARE) {
        state.getCurrentPlayer().setMustPlayCard(true);
      }
      // Setze letzte Aktion
      state.setLastAction(this);
    } else {
      throw new InvalidMoveException("Vorwärtszug um " + this.distance + " Felder ist nicht möglich");
    }
  }

  public int getDistance() {
    return distance;
  }

  private void setDistance(int distance) {
    if (distance <= 0) {
      throw new IllegalArgumentException("distance has to be greater than 0");
    }
    this.distance = distance;
  }

  @Override
  public Advance clone() {
    return new Advance(this.distance, this.order);
  }

  @Override
  public boolean equals(Object o) {
    if(o instanceof Advance) {
      return (this.distance == ((Advance) o).distance);
    }
    return false;
  }

  @Override
  public String toString() {
    return "Advance: distance " + this.distance + " order " + this.order;
  }
}

File: plugin/src/main/kotlin/sc/plugin2025/Advance.kt
package sc.plugin2025

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import sc.plugin2025.util.HuIConstants
import sc.shared.IMoveMistake

/**
 * Ein Vorwärtszug um die angegebene Distanz.
 * - verbrauchte Karotten werden mit k = (distance * (distance + 1)) / 2 berechnet (Gaußsche Summe).
 * - falls der Zug auf einem Hasenfeld endet, müssen auszuführende Hasenkarten mitgegeben werden.
 * - falls der Zug auf einem Marktfeld endet, wird eine zu kaufende Hasenkarte mitgegeben.
 *   Der Wert der Karottentauschkarte spielt dann keine Rolle.
 */
@XStreamAlias(value = "advance")
class Advance(@XStreamAsAttribute val distance: Int, vararg val cards: Card): Move {
    
    override fun perform(state: GameState): IMoveMistake? {
        val player = state.currentPlayer
        val check = state.checkAdvance(distance)
        if(check != null)
            return check
        player.advanceBy(distance)
        
        var lastCard: Card? = null
        var bought = false
        return cards.firstNotNullOfOrNull {
            if(bought)
                return HuIMoveMistake.MUST_BUY_ONE_CARD
            if(state.currentField == Field.MARKET) {
                return@firstNotNullOfOrNull player.consumeCarrots(10) ?: run {
                    bought = true
                    player.addCard(it)
                    null
                }
            }
            if(state.currentField != Field.HARE || lastCard?.moves == false)
                return HuIMoveMistake.CANNOT_PLAY_CARD
            lastCard = it
            it.perform(state)
        } ?: HuIMoveMistake.MUST_BUY_ONE_CARD.takeIf {
            // On Market field and no card bought or just moved there through card
            state.currentField == Field.MARKET && !bought
        } ?: HuIMoveMistake.MUST_PLAY_CARD.takeIf {
            // On Hare field and no card played or just moved there through card
            state.currentField == Field.HARE && lastCard?.moves != false
        }
    }
    
    override fun toString(): String {
        return "Vorwärts um $distance${cards.joinToString(prefix = " mit Karten [", postfix = "]")}"
    }
    
    override fun equals(other: Any?): Boolean {
        if(this === other) return true
        if(other !is Advance) return false
        if(distance != other.distance) return false
        if(!cards.contentEquals(other.cards)) return false
        return true
    }
    
    override fun hashCode(): Int =
        distance + cards.contentHashCode() * HuIConstants.NUM_FIELDS
    
}